src/metadata.rs

```diff
@@ -128,13 +194,237 @@ pub trait Metadata: MajorObject {
         Ok(())
     }
 
+    /// For Datasets this sets the dataset name; normally
+    /// application code should not set the "description" for
+    /// GDALDatasets. For RasterBands it is actually a description
+    /// (if supported) or `""`.
     fn set_description(&mut self, description: &str) -> Result<()> {
-        // For Datasets this sets the dataset name; normally
-        // application code should not set the "description" for
-        // GDALDatasets. For RasterBands it is actually a description
-        // (if supported) or "".
         let c_description = CString::new(description.to_owned())?;
         unsafe { gdal_sys::GDALSetDescription(self.gdal_object_ptr(), c_description.as_ptr()) };
         Ok(())
     }
+
+    /// Get an iterator over metadata entries, across all domains.
+    ///
+    /// # Example
+    ///
+    /// ```rust, no_run
+    /// use gdal::{Dataset, Metadata, MetadataEntry};
+    /// # fn main() -> gdal::errors::Result<()> {
+    /// let dataset = Dataset::open("fixtures/tinymarble.tif")?;
+    /// for MetadataEntry { domain, key, value } in dataset.metadata_iter() {
+    ///   let domain = if domain == "" { "DEFAULT".to_string() } else { domain };
+    ///   println!("{domain}: {key}={value}");
+    /// }
+    /// # Ok(())
+    /// # }
+    /// ```
+    /// Output:
+    /// ```text
+    /// IMAGE_STRUCTURE: INTERLEAVE=PIXEL
+    /// DEFAULT: AREA_OR_POINT=Area
+    /// ...
+    /// DERIVED_SUBDATASETS: DERIVED_SUBDATASET_1_NAME=DERIVED_SUBDATASET:LOGAMPLITUDE:fixtures/tinymarble.tif
+    /// DERIVED_SUBDATASETS: DERIVED_SUBDATASET_1_DESC=log10 of amplitude of input bands from fixtures/tinymarble.tif
+    /// ```
+    fn metadata_iter(&self) -> MetadataIter
+    where
+        Self: Sized,
+    {
+        MetadataIter::new(self)
+    }
+}
+
+/// Standalone metadata entry, as returned by iterator from [`Metadata::metadata_iter`].
+///
+/// Defined by it's parent `domain`, and `key`/`value` pair.
+#[derive(Debug, Clone, Eq, PartialEq, Ord, PartialOrd)]
+pub struct MetadataEntry {
+    pub domain: String,
+    pub key: String,
+    pub value: String,
+}
+
+impl MetadataEntry {
+    pub fn new<D, K, V>(domain: D, key: K, value: V) -> Self
+    where
+        D: Into<String>,
+        K: Into<String>,
+        V: Into<String>,
+    {
+        Self {
+            domain: domain.into(),
+            key: key.into(),
+            value: value.into(),
+        }
+    }
+
+    /// Determine if this entry is from the default domain, which is named `""`.
+    pub fn is_default_domain(&self) -> bool {
+        self.domain.is_empty()
+    }
+}
+
+/// Metadata iterator state
+pub struct MetadataIter<'a> {
+    stream: Box<dyn Iterator<Item = MetadataEntry> + 'a>,
```

Comment:
It doesn't matter too much, but could we avoid the `Box` here, for example with https://doc.rust-lang.org/std/iter/fn.from_fn.html (by moving the code into `metadata_iter`).

Reply:
Wow! I didn't know about `std::iter::from_fn`!!

Reply:
Generators can't come too soon :-).

Reply:
@lnicola Please see comments after `metadata_iter` -> `metadata` comment. I should have added them here.